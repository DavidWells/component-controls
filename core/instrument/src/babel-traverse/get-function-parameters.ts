import { Story, CodeSource, StoryParameters } from '../types';

interface KeyType {
  name: string;
}
interface ASTPropNode {
  name?: string;
  loc: CodeSource;
  properties?: any;
  key?: KeyType;
}
export const extractFunctionParameters = (story: Story) => ({
  ArrowFunctionExpression: (path: any) => {
    const node = path.node;
    if (!story.parameters) {
      story.parameters = [];
    }
    const pushParams = (
      node: ASTPropNode,
      parameters: StoryParameters,
      key?: KeyType,
    ) => {
      const loc = {
        start: {
          column: node.loc.start.column,
          line: node.loc.start.line,
        },
        end: {
          column: node.loc.end.column,
          line: node.loc.end.line,
        },
      };
      if (node.name) {
        parameters.push({
          value: node.name,
          name: key ? key.name : node.name,
          loc,
        });
      } else if (node.properties) {
        const nestedParameters: StoryParameters = [];
        parameters.push({
          value: nestedParameters,
          name: key ? key.name : node.name,
          loc,
        });
        node.properties.forEach(
          ({ value, key }: { value: ASTPropNode; key?: KeyType }) => {
            pushParams(value, nestedParameters, key);
          },
        );
      }
    };
    if (node.params) {
      node.params.forEach(
        (p: { name: string; loc: CodeSource; properties?: any }) => {
          if (story.parameters) {
            pushParams(p, story.parameters);
          }
        },
      );
    }
  },
});
