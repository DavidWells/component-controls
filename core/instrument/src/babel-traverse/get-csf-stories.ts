import { Stories, Story } from '../types';
import traverse from '@babel/traverse';
import { extractFunctionParameters } from './get-function-parameters';

export const extractCSFStories = (stories: Stories) => ({
  ExportNamedDeclaration: (path: any) => {
    const {
      declaration: {
        declarations: [declaration],
      },
    } = path.node;
    const el = declaration.init.body;
    const name = declaration.id.name;
    const story: Story = {
      source: {
        start: {
          column: el.loc.start.column,
          line: el.loc.start.line,
        },
        end: {
          column: el.loc.end.column,
          line: el.loc.end.line,
        },
      },
    };
    traverse(path.node, extractFunctionParameters(story), path.scope, path);
    stories[name] = story;
  },
});
