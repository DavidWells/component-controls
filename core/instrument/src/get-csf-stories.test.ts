import * as parser from '@babel/parser';
import traverse from '@babel/traverse';
import { extractCSFStories } from './babel-traverse/get-csf-stories';
export * from './types';

describe('Extract CSF stories', () => {
  const traverseStories = (code: string) => {
    const ast = parser.parse(code, {
      sourceType: 'module',
      plugins: ['jsx', 'typescript'],
    });
    const stories = {};
    traverse(ast, extractCSFStories(stories));
    return stories;
  };
  it('No parameters', () => {
    expect(
      traverseStories(`
      export const myStory = () => {};
    `),
    ).toMatchSnapshot();
  });

  it('Props parameters', () => {
    expect(
      traverseStories(`
      export const myStory = props => {};
    `),
    ).toMatchSnapshot();
  });
  it('Two parameters', () => {
    expect(
      traverseStories(`
      export const myStory = (props, context) => {};
    `),
    ).toMatchSnapshot();
  });
  it('Typescript', () => {
    expect(
      traverseStories(`
      export const myStory = (props: Properties) => {};
    `),
    ).toMatchSnapshot();
  });

  it('Two levels sub parameters', () => {
    expect(
      traverseStories(`
      export const myStory = ({ name, age }) => {};
    `),
    ).toMatchSnapshot();
  });

  it('Two levels - alias', () => {
    expect(
      traverseStories(`
      export const myStory = ({ name: MyNam, age }) => {};
    `),
    ).toMatchSnapshot();
  });
  it('Three levels - alias', () => {
    expect(
      traverseStories(`
      export const myStory = ({ name: { first, last }, age }) => {};
    `),
    ).toMatchSnapshot();
  });
});
