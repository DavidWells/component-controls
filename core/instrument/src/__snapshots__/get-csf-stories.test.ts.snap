// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extract CSF stories No parameters 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [],
    "source": Object {
      "end": Object {
        "column": 37,
        "line": 2,
      },
      "start": Object {
        "column": 35,
        "line": 2,
      },
    },
  },
}
`;

exports[`Extract CSF stories Props parameters 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 2,
          },
          "start": Object {
            "column": 29,
            "line": 2,
          },
        },
        "name": "props",
        "value": "props",
      },
    ],
    "source": Object {
      "end": Object {
        "column": 40,
        "line": 2,
      },
      "start": Object {
        "column": 38,
        "line": 2,
      },
    },
  },
}
`;

exports[`Extract CSF stories Three levels - alias 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 60,
            "line": 2,
          },
          "start": Object {
            "column": 30,
            "line": 2,
          },
        },
        "name": undefined,
        "value": Array [
          Object {
            "loc": Object {
              "end": Object {
                "column": 53,
                "line": 2,
              },
              "start": Object {
                "column": 38,
                "line": 2,
              },
            },
            "name": "name",
            "value": Array [
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 45,
                    "line": 2,
                  },
                  "start": Object {
                    "column": 40,
                    "line": 2,
                  },
                },
                "name": "first",
                "value": "first",
              },
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 51,
                    "line": 2,
                  },
                  "start": Object {
                    "column": 47,
                    "line": 2,
                  },
                },
                "name": "last",
                "value": "last",
              },
            ],
          },
          Object {
            "loc": Object {
              "end": Object {
                "column": 58,
                "line": 2,
              },
              "start": Object {
                "column": 55,
                "line": 2,
              },
            },
            "name": "age",
            "value": "age",
          },
        ],
      },
    ],
    "source": Object {
      "end": Object {
        "column": 67,
        "line": 2,
      },
      "start": Object {
        "column": 65,
        "line": 2,
      },
    },
  },
}
`;

exports[`Extract CSF stories Two levels - alias 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 50,
            "line": 2,
          },
          "start": Object {
            "column": 30,
            "line": 2,
          },
        },
        "name": undefined,
        "value": Array [
          Object {
            "loc": Object {
              "end": Object {
                "column": 43,
                "line": 2,
              },
              "start": Object {
                "column": 38,
                "line": 2,
              },
            },
            "name": "name",
            "value": "MyNam",
          },
          Object {
            "loc": Object {
              "end": Object {
                "column": 48,
                "line": 2,
              },
              "start": Object {
                "column": 45,
                "line": 2,
              },
            },
            "name": "age",
            "value": "age",
          },
        ],
      },
    ],
    "source": Object {
      "end": Object {
        "column": 57,
        "line": 2,
      },
      "start": Object {
        "column": 55,
        "line": 2,
      },
    },
  },
}
`;

exports[`Extract CSF stories Two levels sub parameters 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 43,
            "line": 2,
          },
          "start": Object {
            "column": 30,
            "line": 2,
          },
        },
        "name": undefined,
        "value": Array [
          Object {
            "loc": Object {
              "end": Object {
                "column": 36,
                "line": 2,
              },
              "start": Object {
                "column": 32,
                "line": 2,
              },
            },
            "name": "name",
            "value": "name",
          },
          Object {
            "loc": Object {
              "end": Object {
                "column": 41,
                "line": 2,
              },
              "start": Object {
                "column": 38,
                "line": 2,
              },
            },
            "name": "age",
            "value": "age",
          },
        ],
      },
    ],
    "source": Object {
      "end": Object {
        "column": 50,
        "line": 2,
      },
      "start": Object {
        "column": 48,
        "line": 2,
      },
    },
  },
}
`;

exports[`Extract CSF stories Two parameters 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 35,
            "line": 2,
          },
          "start": Object {
            "column": 30,
            "line": 2,
          },
        },
        "name": "props",
        "value": "props",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 44,
            "line": 2,
          },
          "start": Object {
            "column": 37,
            "line": 2,
          },
        },
        "name": "context",
        "value": "context",
      },
    ],
    "source": Object {
      "end": Object {
        "column": 51,
        "line": 2,
      },
      "start": Object {
        "column": 49,
        "line": 2,
      },
    },
  },
}
`;

exports[`Extract CSF stories Typescript 1`] = `
Object {
  "myStory": Object {
    "parameters": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 47,
            "line": 2,
          },
          "start": Object {
            "column": 30,
            "line": 2,
          },
        },
        "name": "props",
        "value": "props",
      },
    ],
    "source": Object {
      "end": Object {
        "column": 54,
        "line": 2,
      },
      "start": Object {
        "column": 52,
        "line": 2,
      },
    },
  },
}
`;
